File: backend/app/api/sync.py
"""Manual sync and engine trigger API."""

from typing import Any

from fastapi import APIRouter, Depends
from sqlalchemy import update
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import UserContext, db_session, get_current_user, require_permission
from app.core.permissions import RESTOCK_OPERATE, SYNC_OPERATE, SYNC_VIEW
from app.models.global_config import GlobalConfig
from app.schemas.sync import (
    EngineRunIn,
    SchedulerStatusOut,
    SchedulerToggleIn,
)
from app.tasks.queue import enqueue_task
from app.tasks.scheduler import reload_scheduler, scheduler_status

router = APIRouter(prefix="/api", tags=["sync"])


async def _enqueue(db: AsyncSession, job_name: str) -> dict[str, Any]:
    task_id, existing = await enqueue_task(db, job_name=job_name, trigger_source="manual")
    return {"task_id": task_id, "existing": existing}


@router.post("/sync/all")
async def sync_all(
    db: AsyncSession = Depends(db_session),
    user: UserContext = Depends(get_current_user),
    _: None = Depends(require_permission(SYNC_OPERATE)),
) -> dict[str, Any]:
    return await _enqueue(db, "sync_all")


@router.post("/sync/shop")
async def sync_shop(
    db: AsyncSession = Depends(db_session),
    user: UserContext = Depends(get_current_user),
    _: None = Depends(require_permission(SYNC_OPERATE)),
) -> dict[str, Any]:
    return await _enqueue(db, "sync_shop")


@router.post("/sync/product-listing")
async def sync_product_listing(
    db: AsyncSession = Depends(db_session),
    user: UserContext = Depends(get_current_user),
    _: None = Depends(require_permission(SYNC_OPERATE)),
) -> dict[str, Any]:
    return await _enqueue(db, "sync_product_listing")


@router.post("/sync/inventory")
async def sync_inventory(
    db: AsyncSession = Depends(db_session),
    user: UserContext = Depends(get_current_user),
    _: None = Depends(require_permission(SYNC_OPERATE)),
) -> dict[str, Any]:
    return await _enqueue(db, "sync_inventory")


@router.post("/sync/out-records")
async def sync_out_records(
    db: AsyncSession = Depends(db_session),
    user: UserContext = Depends(get_current_user),
    _: None = Depends(require_permission(SYNC_OPERATE)),
) -> dict[str, Any]:
    return await _enqueue(db, "sync_out_records")


@router.post("/sync/orders")
async def sync_orders(
    db: AsyncSession = Depends(db_session),
    user: UserContext = Depends(get_current_user),
    _: None = Depends(require_permission(SYNC_OPERATE)),
) -> dict[str, Any]:
    return await _enqueue(db, "sync_order_list")


@router.post("/sync/warehouse")
async def sync_warehouse(
    db: AsyncSession = Depends(db_session),
    user: UserContext = Depends(get_current_user),
    _: None = Depends(require_permission(SYNC_OPERATE)),
) -> dict[str, Any]:
    return await _enqueue(db, "sync_warehouse")


@router.get("/sync/scheduler", response_model=SchedulerStatusOut)
async def get_scheduler_status(
    user: UserContext = Depends(get_current_user),
    _: None = Depends(require_permission(SYNC_VIEW)),
) -> SchedulerStatusOut:
    return await scheduler_status()


@router.post("/sync/scheduler", response_model=SchedulerStatusOut)
async def set_scheduler_status(
    payload: SchedulerToggleIn,
    db: AsyncSession = Depends(db_session),
    user: UserContext = Depends(get_current_user),
    _: None = Depends(require_permission(SYNC_OPERATE)),
) -> SchedulerStatusOut:
    await db.execute(
        update(GlobalConfig).where(GlobalConfig.id == 1).values(scheduler_enabled=payload.enabled)
    )
    await db.commit()
    return await reload_scheduler()


@router.post("/engine/run")
async def run_engine_now(
    payload: EngineRunIn,
    db: AsyncSession = Depends(db_session),
    user: UserContext = Depends(get_current_user),
    _: None = Depends(require_permission(RESTOCK_OPERATE)),
) -> dict[str, Any]:
    task_id, existing = await enqueue_task(
        db,
        job_name="calc_engine",
        trigger_source="manual",
        payload={"triggered_by": "manual", "demand_date": payload.demand_date.isoformat()},
    )
    return {"task_id": task_id, "existing": existing}

File: frontend/src/config/sync.ts
export interface SyncActionDefinition {
  key: string
  jobName: string
  label: string
  url: string
}

export interface AutoSyncDefinition {
  jobName: string
  label: string
  cadence: string
}

export const ORDER_DETAIL_REFETCH_JOB_NAME = 'refetch_order_detail'
export const ORDER_DETAIL_REFETCH_LABEL = '详情获取'

export const manualSyncActions: SyncActionDefinition[] = [
  {
    key: 'sync_all',
    jobName: 'sync_all',
    label: '全量同步',
    url: '/api/sync/all',
  },
  {
    key: 'sync_shop',
    jobName: 'sync_shop',
    label: '店铺同步',
    url: '/api/sync/shop',
  },
  {
    key: 'sync_warehouse',
    jobName: 'sync_warehouse',
    label: '仓库同步',
    url: '/api/sync/warehouse',
  },
  {
    key: 'sync_product_listing',
    jobName: 'sync_product_listing',
    label: '商品同步',
    url: '/api/sync/product-listing',
  },
  {
    key: 'sync_inventory',
    jobName: 'sync_inventory',
    label: '库存同步',
    url: '/api/sync/inventory',
  },
  {
    key: 'sync_order_list',
    jobName: 'sync_order_list',
    label: '订单同步',
    url: '/api/sync/orders',
  },
  {
    key: 'sync_out_records',
    jobName: 'sync_out_records',
    label: '出库记录同步',
    url: '/api/sync/out-records',
  },
  {
    key: 'backfill_out_record_target_country',
    jobName: 'backfill_out_record_target_country',
    label: '回填出库目标国家',
    url: '/api/sync/out-records/backfill-target-country',
  },
]

export const replenishmentAction = {
  key: 'engine_run',
  label: '生成补货建议',
  url: '/api/engine/run',
}

export const autoSyncDefinitions: AutoSyncDefinition[] = [
  {
    jobName: 'sync_shop',
    label: '店铺基础同步',
    cadence: '建议每天执行 1 次，或授权状态变更后立即执行。',
  },
  {
    jobName: 'sync_warehouse',
    label: '仓库基础同步',
    cadence: '建议每天执行 1 次，或仓库配置调整后执行。',
  },
  {
    jobName: 'sync_product_listing',
    label: '商品基础同步',
    cadence: '建议每 1-2 小时执行 1 次。',
  },
  {
    jobName: 'sync_inventory',
    label: '库存同步',
    cadence: '建议每 30-60 分钟执行 1 次。',
  },
  {
    jobName: 'sync_order_list',
    label: '订单列表同步',
    cadence: '建议每 30-60 分钟执行 1 次。',
  },
  {
    jobName: 'sync_order_detail',
    label: '订单详情同步',
    cadence: '建议紧跟订单列表同步后执行。',
  },
  {
    jobName: 'sync_out_records',
    label: '出库记录同步',
    cadence: '建议每 2-4 小时执行 1 次。',
  },
]

export const syncJobLabelMap = Object.fromEntries(
  [
    ...manualSyncActions.map((item) => [item.jobName, item.label]),
    ...autoSyncDefinitions.map((item) => [item.jobName, item.label]),
    [ORDER_DETAIL_REFETCH_JOB_NAME, ORDER_DETAIL_REFETCH_LABEL],
    ['calc_engine', replenishmentAction.label],
    ['daily_archive', '每日归档'],
  ] as const,
)
